### DISCUSSION TITLE
VipsJpeg: Wrong JPEG library version: library is 62, caller expects 80

### DISCUSSION THREAD

**Discussion by xfuturomax** [libvips/libvips#2829] [Help wanted]
I use Debian 9 (scratch), and this OS has only 8.4.5-1 version for vips in packages. This version has some bugs that blocks image processing. So I decided to build the latest version from sources.
After building it and running I got an error: `VipsJpeg: Wrong JPEG library version: library is 62, caller expects 80`.

Deb9 has only libjpeg62-turbo in packages, so I built 8d version from sources and now there are two versions of jpeglib in the system. I can't remove old version because many libraries depend on it. 

And then I tried to configure vips with arguments:
`./configure --with-jpeg-includes=/usr/local/include --with-jpeg-libraries=/usr/local/lib`, because 8 version installed in /usr/local.
But it didn't help. I still get the same error.

It would be great if somebody could help me.

**Comment by jcupitt** [comment]
Hi @xfuturomax,

It should work with the libjpeg in debian9. I'd remove the one you built and try again. 

This dockerfile will build current stable libvips for debian9:

```Dockerfile
FROM debian:9.11

RUN apt-get update && apt-get install -y \
        build-essential \
        wget \
        pkg-config

# stuff we need to build our own libvips ... this is a pretty random selection
# of dependencies, you'll want to adjust these
RUN apt-get install -y \
        glib-2.0-dev \
        libexpat-dev \
        librsvg2-dev \
        libpng-dev \
        libgif-dev \
        libjpeg-dev \
        libexif-dev \
        liblcms2-dev \
        liborc-dev

ARG VIPS_VERSION=8.12.2
ARG VIPS_URL=https://github.com/libvips/libvips/releases/download

WORKDIR /usr/local/src

RUN wget ${VIPS_URL}/v${VIPS_VERSION}/vips-${VIPS_VERSION}.tar.gz \
        && tar xf vips-${VIPS_VERSION}.tar.gz \
        && cd vips-${VIPS_VERSION} \
        && ./configure \
        && make \
        && make install
```

I can run it like this:

```
$ docker build -t php-vips-debian9 .
$ docker run -it --rm -v $PWD:/data php-vips-debian9
root@e04ba2701910:/usr/local/src# vipsheader /data/k2.jpg
/data/k2.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
```

You can build your own libjpeg, but you'd also need to build everything that libvips depends on that indirectly depends on the system libjpeg, and that's a lot of packages.

**Comment by xfuturomax** [comment]
Hi @jcupitt 

Thank you for your answer. 

At the moment I don't use docker on these server for certain reasons. And I also need to use vips inside another library so I think docker doesn't suit here.

As I mentioned I've already built vips without any problem, the problem only is that it doesn't use 8 version of libjpeg, even if I specify it during configuration stage. How can I adjust vips to use 8 version or remove this error?

I've also tried to link 62 version to 8, yes It's a dirty trick, but anyway it also didn't work.

When I'm trying to remove 62 I see warning about dependencies: dpkg: dependency problems prevent removal of libjpeg62-turbo:amd64:.

**Comment by jcupitt** [comment]
I didn't mean you should use docker, I was trying to show that you can build current libvips on debian9 without needing a new JPEG library.

Changing your libjpeg library is difficult: it's used everywhere, so you need to rebuild all the packages which use it, and there are a lot of them.

I would remove your libjpeg v8 and stick to the system libjpeg, unless there is a very strong reason to change.

**Comment by xfuturomax** [comment]
I've removed v8 of libjpeg, and rebuilt vips, but unfortunately it didn't' help.
This is the details from ldconfig:

>$ sudo ldconfig -v | grep jpeg
	libmjpegutils-2.1.so.0 -> libmjpegutils-2.1.so.0.0.0
	libjpegxr.so.0 -> libjpegxr.so.1.1
	libjpeg.so.62 -> libjpeg.so.62.2.0
	libturbojpeg.so.0 -> libturbojpeg.so.0.1.0

>$ sudo ldconfig -v | grep vips
	libvips.so.42 -> libvips.so.42.14.2
	libvips-cpp.so.42 -> libvips-cpp.so.42.14.2

And the error:

>$ vipsheader ~dev/fly-images/LOGO_01.jpg
vipsheader: VipsJpeg: Wrong JPEG library version: library is 62, caller expects 80

**Reply by jcupitt** [reply]
The message means that libvips has been built against libjpeg 8 headers, but is seeing the libjpeg 6.2 library at runtime.

I think you must have some libjpeg v8 headers still on your system. I'd check `pkg-config` and maybe have a look in `/usr/local`.

**Comment by xfuturomax** [comment]
Thanks a lot, I'll try to investigate this case. Another question, is there a way to strictly specify path to jpeglib to built libvips? because it looks like this script is not used by the build script:
./configure --with-jpeg-includes=/usr/local/include --c=/usr/local/lib

**Reply by jcupitt** [reply]
I think the `--with-jpeg-includes` is only used if pkg-config fails to find your libjpeg. Try:

```
pkg-config libjpeg --cflags --libs
```

And see what it's set up as.

**Reply by xfuturomax** [reply]
It prints just -ljpeg .

**Reply by jcupitt** [reply]
Ah, then you probably have jpeg8 headers in /usr/include and (possibly?) a jpeg8 libvips.pc in /usr/lib/pkgconfig. I'd check and clean them out.

**Reply by xfuturomax** [reply]
You were right, there was 80 version here: /usr/include/jconfig.h, so I changed it to 62 here: `define JPEG_LIB_VERSION 62` and then reconfigured and rebuilt vips, and thats all! Thank you so much for your help and patience!
